### ISSUE TITLE
Storage-Plugin should support async storage engines

### ISSUE THREAD

**Issue by BeaveArony** [ngxs/store#221]
Some storage engines are promise or callback based. Like LocalForage, idb-keyval or native mobile storage implementations.

Probably the main challenge is to delay the first rendering of the app until the state is actually rehydrated from an async storage. redux-persist does this with react.
I dunno, but maybe ngxs could hook into the router to cancel the first navigation attempt, waiting for some callback from the plugin to redo the navigation once rehydration is completed? But not every app uses the router ...

Another useful feature would be to define blacklisted keys. Imagine you have a big feature state you want to persist, except one key (ie. loading or pending). You would have to define a huge list of keys to make it happen, right?

**Comment by AlgoTrader** [comment]
Render is probably not the state management business. App should decide itself how to do it. If you have async state, it is better to provide default state immediatly and then patch it with real state forcing re-render

**Comment by amcdnl** [comment]
I think the vast majority of cases will want sync. I imagine you have a indexdb use case or something. It would be easy to make a new plugin to support this but not something that is priority on my list right now.

**Comment by BeaveArony** [comment]
Thanks for the answers.
I don't know, if the vast majority will (want) localStorage. It's slow (you need to JSON.stringify()/parse() everything), blocks the UI-Thread and is very limited with mostly 5MB of data and is only supported in browsers. As soon as you want to use NativeScript or IndexedDB or whatever comes in the future, you will most likely need async and the ability to react on the rehydration event.

I don't think that NGXS should bring all of those features by default, but since it provides quite some abstraction of redux and the bootstrapping process, it should offer some kind of entry into the initialization point of the store, so we could wait on some callback to render our app for example. Maybe I'm missing something and it's easily doable?

Plugins are comparable to middleware in redux. You hook into each action to do something before and/or after and you might even swallow it. Is it possible to hook into the createStore() process to alter the initial state from within a plugin?

**Comment by amcdnl** [comment]
Closing and marking for future.

**Comment by alejandro-heredia-coderoad-com** [comment]
Couldn't agree more with @BeaveArony, NGXS is just too powerful, smart and stable to only support sync storage. Async storage engines should be priority.

**Comment by kctang** [comment]
I am using NGXS in an Ionic application that needs to support various devices. LocalStorage support in older devices is quite 'unpredictable' to say the least (even iOS 10.3.3)! So, having the option to swap storage engine with things like Sqlite or IndexDB will be critical.

Any chance we can up the priority of this item?

**Comment by johngrimsey** [comment]
+100000 on this. 

@amcdnl All real-world usage examples I can think of are async. For example, PouchDB returns promises, Angular's own HttpClient returns an observable. 

Might you reconsider?

**Comment by splincode** [comment]
@markwhitfeld ping

**Comment by ralphbert** [comment]
+1

**Comment by TomDemulierChevret** [comment]
I've made some test and managed to use ``@ngxs/storage-plugin`` with [Ionic storage service](https://beta.ionicframework.com/docs/building/storage/) (which uses internally https://github.com/localForage/localForage).

I'll make a PR tomorrow.

**Comment by arturovt** [comment]
Closing this in favour of [async-storage-plugin](https://github.com/ngxs-labs/async-storage-plugin)

**Comment by stephan-fischer** [comment]
@arturovt  Could you please open this issue? 
[async-storage-plugin](https://github.com/ngxs-labs/async-storage-plugin) seems to be no longer maintained, and do not work for newer angular versions.
